File: general-interview-prep/Practice_Problems_Google_Focused/isValidBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
const isValidBST = (root, min = -Infinity, max = Infinity) => {
    if (root === null) {
        return true;
    }

    // false if this node violates min/max constraint
    if (root.val < min || root.val > max) {
        return false;
    }

    // Otherwise check the subtrees recursively
    // tightening the min or max constraint
    return (isValidBST(root.left, min, root.val -1) && isValidBST(root.right, root.val+1, max))
};

File: general-interview-prep/golang/isValidBST.go
package main

import "fmt"

type TreeNode struct {
   Val int
   Left *TreeNode
   Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
    if root == nil {
        return true
    }

    min := math.MinInt64
    max := math.MaxInt64

    return recurse(root.Left, min, root.Val - 1) && recurse(root.Right, root.Val + 1, max)
}

func recurse(root *TreeNode, min, max int) bool {
    if root == nil {
        return true
    }
    if root.Val < min || root.Val > max {
        return false
    }

    return recurse(root.Left, min, root.Val - 1) && recurse(root.Right, root.Val + 1, max)
}


func main() {

}
